akka-projection-jdbc/src/test/scala/akka/projection/jdbc/JdbcOffsetStoreSpec.scala

```diff
@@ -0,0 +1,281 @@
+import akka.projection.jdbc.javadsl.JdbcSession.tryWithResource
+import akka.projection.jdbc.javadsl.JdbcSession.withConnection
+import akka.projection.testkit.internal.TestClock
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import org.scalatest.OptionValues
+import org.scalatest.time.Millis
+import org.scalatest.time.Seconds
+import org.scalatest.time.Span
+
+object JdbcOffsetStoreSpec {
+
+  trait JdbcSpecConfig {
+    val name: String
+    val config: Config
+    val jdbcSessionFactory: () => JdbcSession
+  }
+
+  private[akka] class PureJdbcSession(connFunc: () => Connection) extends JdbcSession {
+    lazy val conn = connFunc()
+    override def withConnection[Result](func: function.Function[Connection, Result]): Result =
+      func(conn)
+
+    override def commit(): Unit = conn.commit()
+
+    override def rollback(): Unit = conn.rollback()
+
+    override def close(): Unit = conn.close()
+  }
+
+  object H2SpecConfig extends JdbcSpecConfig {
+
+    val name = "H2 Database"
+    val config: Config = ConfigFactory.parseString("""
+    akka.projection.jdbc = {
+      dialect = "h2-dialect"
+      offset-store {
+        schema = ""
+        table = "AKKA_PROJECTION_OFFSET_STORE"
+      }
+      
+      # TODO: configure a connection pool for the tests
+      blocking-jdbc-dispatcher.thread-pool-executor.fixed-pool-size = 5
+    }
+    """)
+
+    val jdbcSessionFactory = () =>
+      new PureJdbcSession(() => {
+        Class.forName("org.h2.Driver")
+        val c = DriverManager.getConnection("jdbc:h2:mem:offset-store-test;DB_CLOSE_DELAY=-1")
+        c.setAutoCommit(false)
+        c
+      })
+  }
+}
+
+abstract class JdbcOffsetStoreSpec(specConfig: JdbcSpecConfig) extends JdbcSpec(specConfig.config) with OptionValues {
+
+  override implicit val patienceConfig: PatienceConfig =
+    PatienceConfig(timeout = Span(3, Seconds), interval = Span(100, Millis))
+
+  implicit val executionContext = testKit.system.executionContext
+
+  // test clock for testing of the `last_updated` Instant
+  private val clock = new TestClock
+
+  private val settings = JdbcSettings(testKit.system)
+  private val offsetStore = new JdbcOffsetStore(settings, specConfig.jdbcSessionFactory, clock)
+  private val dialectLabel = specConfig.name
+
+  override protected def beforeAll(): Unit = {
+    // create offset table
+    Await.result(offsetStore.createIfNotExists(), 3.seconds)
+  }
+
+  override protected def afterAll(): Unit = {}
+
+  private def selectLastUpdated(projectionId: ProjectionId): Instant = {
+    withConnection(specConfig.jdbcSessionFactory) { conn =>
+      val statement = s"SELECT * FROM ${settings.table} WHERE projection_name = ? AND projection_key = ?"
+      tryWithResource(conn.prepareStatement(statement)) { stmt =>
+        stmt.setString(1, projectionId.name)
+        stmt.setString(2, projectionId.key)
+        val resultSet = stmt.executeQuery()
+
+        if (resultSet.first()) {
```

Comment:
I think `resultSet` must also be closed (at least that is what I remember from back in the days)